script.js

```diff
@@ -1 +1,464 @@
+  if (timeoutIntervalReference !== null)
+  {
+    clearInterval(timeoutIntervalReference);
+  }
+  console.log(`Before: ${boardOfCards}`);
+  boardOfCards.length = 0;
+  console.log(`After: ${boardOfCards}`);
+  if (boardElement !== null)
+  {
+    boardElement.innerHTML = '';
+  }
+  divGameStatusInfo.innerHTML = '';
+  // setTimerMessage('');
+  gameTotalScore = 0;
+  gameAttempts = 0;
+  gameStarted = false;
+};
+
+// This function performs all other intializations of the game
+const startGame = () => {
+  console.log('Start Game');
+  // If the game is already started
+  if (gameStarted)
+  {
+    // For handling the second automatic start, before the intervelled call to the game happens
+    // from the function onClickStartButton
+    resetElements();
+  }
+  gameStarted = true;
+  // create this special deck by getting the doubled cards and
+  // making a smaller array that is ( boardSize squared ) number of cards
+  const doubleDeck = makeDeck();
+  const deckSubset = doubleDeck.slice(0, boardSize * boardSize);
+  deck = shuffleCards(deckSubset);
+
+  // deal the cards out to the board data structure
+  for (let i = 0; i < boardSize; i += 1) {
+    boardOfCards.push([]);
+    for (let j = 0; j < boardSize; j += 1) {
+      boardOfCards[i].push(deck.pop());
+    }
+  }
+  boardElement = buildBoardElements(boardOfCards);
+  document.body.appendChild(boardElement);
+};
+
+// Function to display the player Name
+const displayPlayerName = () => {
+  playerName = inputPlayerName.value;
+  setGameStatusInfo(`Hey ${playerName}, You can start playing.`);
+};
+
+const playGameWithTimer = () => {
+  if (playerName === '')
+  {
+    setGameStatusInfo('Please enter your name to start playing.');
+    return;
+  }
+  countDownTime = intervalInMilliSeconds * 5;
+  setTimerMessage(`Time remaining: ${convertMillisecondsToSeconds(countDownTime)} seconds.`);
+  startGame();
+
+  timeoutIntervalReference = setInterval(() => {
+    countDownTime -= intervalSingleSecond; // decrease the counter by interval time for the timer
+    setTimerMessage(`Time remaining: ${convertMillisecondsToSeconds(countDownTime)} seconds.`);
+    const playerWin = doesPlayerWin();
+    if (playerWin || countDownTime <= 0) // If the counter reaches zero, restart the game
+    {
+      if (playerWin)
+      {
+        setTimerMessage(`${playerName} wins. Please press Start if you would like to play more.`);
+        countDownTime = 0;
+      }
+      else { setTimerMessage(`${playerName} lost. Time over. Stopping the game.`); }
+      resetElements();
+      clearInterval(timeoutIntervalReference);
+    }
+  }, intervalSingleSecond);
+};
+
+// Function that starts the game. Here it resets the board of cards
+// Also it sets an interval to restart the game after a specific amount of time
+const onClickStartButton = () => {
+  console.log('onClickStart');
+  // startGame();
+  // intervalReference = setInterval(startGame, intervalInMilliSeconds);
+
+  playGameWithTimer();
+};
```

Comment:
Great decomposition with helper functions to abstract complexity into smaller chunks!